processing/src/main/java/org/apache/druid/segment/join/JoinableFactoryWrapper.java

```diff
@@ -144,9 +144,15 @@ static JoinClauseToFilterConversion convertJoinToFilter(
       final Set<String> rightPrefixes
   )
   {
+    // This optimization kicks in when there is exactly 1 equijoin
+    // The reason being that getNonNullColumnValues uses a TreeSet for handling duplicates
+    // and does not return values in order.
+    // In case TreeSet if replaced by a LinkedHashSet, duplicates are not handled correctly
+    // and resulting ordering is improper.
+    // Considering these joins are converted to filters only when there is 1 equijoin
     if (clause.getJoinType() == JoinType.INNER
         && clause.getCondition().getNonEquiConditions().isEmpty()
-        && clause.getCondition().getEquiConditions().size() > 0) {
+        && clause.getCondition().getEquiConditions().size() == 1) {
       final List<Filter> filters = new ArrayList<>();
```

Comment:
No need for `filters` anymore, as there is always going to be exactly one. You can use `CollectionUtils.getOnlyElement` to get the one condition and process it directly. It will simplify the later logic.

Reply:
Yup, no need for the for loop anymore. Made things simpler